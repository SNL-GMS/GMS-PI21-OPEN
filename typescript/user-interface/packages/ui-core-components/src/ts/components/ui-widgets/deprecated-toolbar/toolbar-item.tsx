/* eslint-disable react/destructuring-assignment */
import { UILogger } from '@gms/ui-util';
import React from 'react';

import type { PopoverButton } from '../popover-button';
// !FIX ESLINT CIRCULAR DEPENDENCY
// eslint-disable-next-line import/no-cycle
import * as Renderers from './toolbar-item-renderers';
import type {
  ButtonGroupItem,
  ButtonItem,
  CheckboxDropdownItem,
  CustomItem,
  DateRangePickerItem,
  DeprecatedToolbarItemBase,
  DropdownItem,
  IntervalPickerItem,
  LabelValueItem,
  LoadingSpinnerItem,
  NumericInputItem,
  PopoverItem,
  SwitchItem,
  ToolbarItem
} from './types';
import { ToolbarItemType } from './types';

const logger = UILogger.create('GMS_LOG_TOOLBAR', process.env.GMS_LOG_TOOLBAR);

export interface ToolbarItemRendererProps {
  item: ToolbarItem;
  hasIssue: boolean;
  addToPopoverMap(rank: number, ref: PopoverButton): void;
}

/**
 * Selects the correct function to render a toolbar item
 */
// eslint-disable-next-line complexity
export function ToolbarItemRenderer(props: ToolbarItemRendererProps): JSX.Element {
  const itemTypes = ToolbarItemType;
  switch ((props.item as DeprecatedToolbarItemBase).type) {
    case itemTypes.Dropdown: {
      return Renderers.renderDropdown(props.item as DropdownItem);
    }
    case itemTypes.NumericInput: {
      return Renderers.renderNumeric(props.item as NumericInputItem);
    }
    case itemTypes.IntervalPicker: {
      return Renderers.renderIntervalPicker(props.item as IntervalPickerItem);
    }
    case itemTypes.DateRangePicker: {
      return Renderers.renderDateRangePicker(props.item as DateRangePickerItem);
    }
    case itemTypes.Popover: {
      return Renderers.renderPopoverButton(props.item as PopoverItem, (key, val) =>
        props.addToPopoverMap(key, val)
      );
    }
    case itemTypes.Switch: {
      return Renderers.renderSwitch(props.item as SwitchItem);
    }
    case itemTypes.Button: {
      return Renderers.renderButton(props.item as ButtonItem);
    }
    case itemTypes.ButtonGroup: {
      return Renderers.renderButtonGroup(props.item as ButtonGroupItem);
    }
    case itemTypes.LabelValue: {
      return Renderers.renderLabelValue(props.item as LabelValueItem, props.hasIssue);
    }
    case itemTypes.CheckboxList:
      return Renderers.renderCheckboxDropdown(props.item as CheckboxDropdownItem, (key, val) =>
        props.addToPopoverMap(key, val)
      );
    case itemTypes.LoadingSpinner:
      return Renderers.renderLoadingSpinner(props.item as LoadingSpinnerItem);
    case itemTypes.CustomItem:
      return (props.item as CustomItem).element;
    default: {
      logger.warn(`default error`);
      return <div />;
    }
  }
}
