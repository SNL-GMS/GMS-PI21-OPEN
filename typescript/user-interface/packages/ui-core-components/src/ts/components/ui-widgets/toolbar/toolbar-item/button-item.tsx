import { Alignment, Button, Icon } from '@blueprintjs/core';
import React from 'react';

import type { ToolbarItemBase, ToolbarItemElement } from '../types';

/**
 * Type guard, for use when rendering overflow menu items.
 */
export function isButtonToolbarItem(object: unknown): object is ButtonToolbarItemProps {
  return (object as ButtonToolbarItemProps).onButtonClick !== undefined;
}

/**
 * Properties to pass to the {@link ButtonToolbarItem}
 *
 * @see {@link ToolbarItemBase} for base properties.
 */
export interface ButtonToolbarItemProps extends ToolbarItemBase {
  onButtonClick();
  marginRight?: number;
}

/**
 * Represents a group of buttons used within a toolbar
 *
 * @param buttonItem the buttonItem to display {@link ButtonItem}
 */
// eslint-disable-next-line complexity, react/function-component-definition
export const ButtonToolbarItem: React.FC<ButtonToolbarItemProps> = ({
  marginRight = 0,
  widthPx,
  onlyShowIcon,
  style,
  disabled,
  label,
  labelRight,
  tooltip,
  icon,
  onButtonClick,
  onMouseEnter,
  onMouseOut,
  cyData
}: ButtonToolbarItemProps): ToolbarItemElement => {
  const widthAsString = widthPx ? `${widthPx}px` : undefined;
  const width = widthAsString || (onlyShowIcon ? '30px' : undefined);

  return (
    <div style={style ?? {}}>
      <Button
        className={onlyShowIcon ? 'toolbar-button--icon-only' : 'toolbar-button'}
        style={{
          width,
          marginRight: marginRight ? `${marginRight}px` : undefined
        }}
        disabled={disabled}
        alignText={onlyShowIcon && !labelRight ? Alignment.CENTER : Alignment.LEFT}
        onClick={onButtonClick}
        title={tooltip}
        onMouseEnter={onMouseEnter}
        onMouseOut={onMouseOut}
        data-cy={cyData}
      >
        {onlyShowIcon ? undefined : <span key="toolbar-button--label">{label}</span>}
        {icon ? <Icon icon={icon} title={false} /> : null}
        {!labelRight ? undefined : <span key="toolbar-button--label-right">{labelRight}</span>}
      </Button>
    </div>
  );
};
