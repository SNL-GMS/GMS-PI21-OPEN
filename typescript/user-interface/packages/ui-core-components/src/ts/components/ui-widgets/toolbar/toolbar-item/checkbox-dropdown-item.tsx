import type Immutable from 'immutable';
import React from 'react';

import { CheckboxList } from '../../checkbox-list';
import { PopoverButton } from '../../popover-button';
import type { ToolbarItemBase, ToolbarItemElement } from '../types';

/**
 * Type guard, for use when rendering overflow menu items.
 */
export function isCheckboxDropdownToolbarItem(
  object: unknown
): object is CheckboxDropdownToolbarItemProps {
  return (object as CheckboxDropdownToolbarItemProps).enumOfKeys !== undefined;
}

/**
 * props for {@link CheckboxDropdownToolbarItem}
 *
 * @see {@link ToolbarItemBase} for base properties.
 */
export interface CheckboxDropdownToolbarItemProps extends ToolbarItemBase {
  /** map for checkbox values (checked/unchecked) */
  values: Immutable.Map<any, boolean>;

  /** map of background colors */
  colors?: Immutable.Map<any, string>;

  /** map of keys to display strings (labels) */
  enumKeysToDisplayStrings?: Immutable.Map<string, string>;

  enumKeysToRenderDividers?: Immutable.Map<string, boolean>;

  enumKeysToLabelStrings?: Immutable.Map<string, string>;

  /** keys */
  enumOfKeys: any;

  /** callback to change event */
  onChange(value: any): void;

  /** callback to onPopUp (list appearing) event */
  onPopUp?(ref?: HTMLDivElement): void;

  /** callback to onPopoverDismissed (list disappears) event */
  onPopoverDismissed?(): void;
}

/**
 * Represents a dropdown list of checkbox items used within a toolbar
 *
 * @param checkboxItem the checkboxItem to display {@link CheckboxDropdownItem}
 */
// eslint-disable-next-line react/function-component-definition
export const CheckboxDropdownToolbarItem: React.FC<CheckboxDropdownToolbarItemProps> = ({
  values,
  colors,
  enumKeysToDisplayStrings,
  enumKeysToLabelStrings,
  enumKeysToRenderDividers,
  enumOfKeys,
  onChange,
  onPopUp,
  onPopoverDismissed,
  style,
  label,
  tooltip,
  disabled,
  widthPx,
  popoverButtonMap,
  cyData
}: CheckboxDropdownToolbarItemProps): ToolbarItemElement => {
  const handleRef = (ref: PopoverButton, buttonMap: Map<number, PopoverButton>): void => {
    if (ref && buttonMap) {
      buttonMap.set(1, ref);
    }
  };

  return (
    <div style={style ?? {}}>
      <PopoverButton
        label={label}
        tooltip={tooltip}
        disabled={disabled}
        cyData={cyData}
        popupContent={
          <CheckboxList
            enumToCheckedMap={values}
            enumToColorMap={colors}
            checkboxEnum={enumOfKeys}
            enumKeysToDisplayStrings={enumKeysToDisplayStrings}
            enumKeysToDividerMap={enumKeysToRenderDividers}
            enumKeysToLabelMap={enumKeysToLabelStrings}
            onChange={onChange}
          />
        }
        onPopoverDismissed={onPopoverDismissed}
        widthPx={widthPx}
        onClick={onPopUp}
        ref={ref => handleRef(ref, popoverButtonMap)}
      />
    </div>
  );
};
