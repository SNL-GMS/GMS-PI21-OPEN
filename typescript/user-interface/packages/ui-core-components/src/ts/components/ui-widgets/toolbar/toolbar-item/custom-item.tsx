import type React from 'react';

import type { ToolbarItemBase, ToolbarItemElement } from '../types';

/**
 * Type guard, for use when rendering overflow menu items.
 */
export function isCustomToolbarItem(object: unknown): object is CustomItemProps {
  return (object as CustomItemProps).element !== undefined;
}

/**
 * type for the custom item toolbar item
 */
export interface CustomItemProps extends ToolbarItemBase {
  element: JSX.Element;
}

export const CustomToolbarItem: React.FC<CustomItemProps> = ({
  element
}: CustomItemProps): ToolbarItemElement => {
  return element;
};
