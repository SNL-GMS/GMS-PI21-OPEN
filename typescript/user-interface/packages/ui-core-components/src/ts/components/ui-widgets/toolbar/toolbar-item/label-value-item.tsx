import React from 'react';

import { LabelValue } from '../../label-value';
import type { ToolbarItemBase, ToolbarItemElement } from '../types';

/**
 * Type guard, for use when rendering overflow menu items.
 */
export function isLabelValueToolbarItem(object: unknown): object is LabelValueToolbarItemProps {
  return (object as LabelValueToolbarItemProps).labelValue !== undefined;
}

/**
 * Properties to pass to the {@link LabelValueToolbarItem}
 *
 * @see {@link ToolbarItemBase} for base properties.
 */
export interface LabelValueToolbarItemProps extends ToolbarItemBase {
  labelValue: string | JSX.Element;
  labelValueColor?: string;
  styleForLabelValue?: React.CSSProperties;
}

/**
 * Represents a label/value pair used within a toolbar
 *
 * @param labelValueItem the item to display {@link LabelValueItem}
 */
// eslint-disable-next-line react/function-component-definition
export const LabelValueToolbarItem: React.FC<LabelValueToolbarItemProps> = ({
  labelValue,
  labelValueColor,
  styleForLabelValue,
  style,
  label,
  ianApp,
  hasIssue,
  tooltip,
  tooltipForIssue,
  cyData
}: LabelValueToolbarItemProps): ToolbarItemElement => {
  return (
    <div className="toolbar-label-value" style={style ?? {}}>
      <LabelValue
        label={label}
        value={labelValue}
        ianApp={ianApp}
        tooltip={hasIssue && tooltipForIssue ? tooltipForIssue : tooltip}
        styleForValue={styleForLabelValue}
        valueColor={labelValueColor}
        data-cy={cyData}
      />
    </div>
  );
};
