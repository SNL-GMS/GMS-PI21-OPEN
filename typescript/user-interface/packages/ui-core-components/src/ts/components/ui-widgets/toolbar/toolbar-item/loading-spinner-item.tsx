import React from 'react';

import { LoadingSpinner } from '../../loading-spinner';
import type { ToolbarItemBase, ToolbarItemElement } from '../types';

/**
 * Type guard, for use when rendering overflow menu items.
 */
export function isLoadingSpinnerToolbarItem(
  object: unknown
): object is LoadingSpinnerToolbarItemProps {
  return (object as LoadingSpinnerToolbarItemProps).itemsToLoad !== undefined;
}

/**
 * Properties to pass to the {@link LoadingSpinnerToolbarItem}
 *
 * @see {@link ToolbarItemBase} for base properties.
 */
export interface LoadingSpinnerToolbarItemProps extends ToolbarItemBase {
  /** count of items to load */
  itemsToLoad: number;

  /** number of items currently loaded */
  itemsLoaded?: number;

  /** hide the label 'Loading' */
  hideTheWordLoading?: boolean;

  /** hid the number of items loaded */
  hideOutstandingCount?: boolean;
}

/**
 * Represents an item loading spinner used within a toolbar with optional counts
 *
 * @param loadingSpinnerItem the item to display {@link LoadingSpinnerItem}
 */
// eslint-disable-next-line react/function-component-definition
export const LoadingSpinnerToolbarItem: React.FC<LoadingSpinnerToolbarItemProps> = ({
  itemsToLoad,
  itemsLoaded,
  hideTheWordLoading,
  hideOutstandingCount,
  style,
  onlyShowIcon,
  label,
  widthPx,
  cyData
}: LoadingSpinnerToolbarItemProps): ToolbarItemElement => {
  return (
    <div style={style ?? {}}>
      <LoadingSpinner
        hideTheWordLoading={hideTheWordLoading}
        hideOutstandingCount={hideOutstandingCount}
        onlyShowSpinner={onlyShowIcon ?? false}
        itemsLoaded={itemsLoaded}
        itemsToLoad={itemsToLoad}
        label={label}
        widthPx={widthPx}
        data-cy={cyData}
      />
    </div>
  );
};
