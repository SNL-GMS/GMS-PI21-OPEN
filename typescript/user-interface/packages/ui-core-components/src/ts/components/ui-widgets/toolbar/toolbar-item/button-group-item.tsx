import React from 'react';

import type { ToolbarItemBase, ToolbarItemElement } from '../types';
import type { ButtonToolbarItemProps } from './button-item';
import { ButtonToolbarItem } from './button-item';

/**
 * Type guard, for use when rendering overflow menu items.
 */
export function isButtonGroupToolbarItem(object: unknown): object is ButtonGroupToolbarProps {
  return (object as ButtonGroupToolbarProps).buttons !== undefined;
}

/**
 * Properties to pass to the {@link ButtonGroupToolbarItem}
 *
 * @see {@link ToolbarItemBase} for base properties.
 */
export interface ButtonGroupToolbarProps extends ToolbarItemBase {
  /** Array of props for {@link ButtonToolbarItem} items to be rendered. */
  buttons: ButtonToolbarItemProps[];
}

/**
 * Represents a group of buttons used within a toolbar
 *
 * @param groupItem Properties to pass to the ButtonGroup {@link ButtonGroupToolbarProps}
 */
// eslint-disable-next-line react/function-component-definition
export const ButtonGroupToolbarItem: React.FC<ButtonGroupToolbarProps> = ({
  buttons,
  style,
  cyData
}: ButtonGroupToolbarProps): ToolbarItemElement => {
  const indexOfLastButton = buttons.length - 1;
  return (
    <div className="toolbar-button-group" data-cy={cyData} style={style ?? {}}>
      {buttons.map((button, index) => {
        return (
          <ButtonToolbarItem
            key={button.key}
            widthPx={button.widthPx}
            onlyShowIcon={button.onlyShowIcon}
            style={button.style}
            disabled={button.disabled}
            label={button.label}
            labelRight={button.labelRight}
            tooltip={button.tooltip}
            icon={button.icon}
            onButtonClick={() => button.onButtonClick()}
            onMouseEnter={button.onMouseEnter ? () => button.onMouseEnter() : undefined}
            onMouseOut={button.onMouseOut ? () => button.onMouseOut() : undefined}
            marginRight={index !== indexOfLastButton ? 2 : 0}
            cyData={button.cyData}
          />
        );
      })}
    </div>
  );
};
