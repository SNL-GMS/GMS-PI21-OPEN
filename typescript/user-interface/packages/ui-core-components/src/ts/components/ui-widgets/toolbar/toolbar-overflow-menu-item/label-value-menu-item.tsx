import { MenuItem } from '@blueprintjs/core';
import React from 'react';

import { isLabelValueToolbarItem } from '../toolbar-item/label-value-item';
import type { ToolbarOverflowMenuItemProps } from './types';

/**
 * ToolbarItem component for a LabelValue specifically in the overflow menu.
 */
// eslint-disable-next-line react/function-component-definition
export const LabelValueOverflowMenuToolbarItem: React.FC<ToolbarOverflowMenuItemProps> = ({
  item,
  menuKey
}: ToolbarOverflowMenuItemProps) => {
  if (isLabelValueToolbarItem(item)) {
    const labelText = item.label && item.label.length > 0 ? `${item.label}: ` : '';
    const menuLabelText =
      typeof item.labelValue === 'string' ? (
        `${labelText} ${item.labelValue}`
      ) : (
        <>
          <span>{labelText}</span>
          {item.labelValue}
        </>
      );
    return (
      <MenuItem
        className={item.hasIssue ? 'toolbar-item--issue' : ''}
        title={item.hasIssue && item.tooltipForIssue ? item.tooltipForIssue : item.tooltip}
        key={menuKey}
        text={menuLabelText}
        disabled={item.disabled}
      />
    );
  }
  return null;
};
