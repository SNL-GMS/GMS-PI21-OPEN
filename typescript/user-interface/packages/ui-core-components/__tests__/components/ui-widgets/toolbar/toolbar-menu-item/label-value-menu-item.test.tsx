import { render } from '@testing-library/react';
import React from 'react';

import { LabelValueToolbarItem } from '../../../../../src/ts/components/ui-widgets/toolbar/toolbar-item/label-value-item';
import { LoadingSpinnerToolbarItem } from '../../../../../src/ts/components/ui-widgets/toolbar/toolbar-item/loading-spinner-item';
import { LabelValueOverflowMenuToolbarItem } from '../../../../../src/ts/components/ui-widgets/toolbar/toolbar-overflow-menu-item/label-value-menu-item';

// set up window alert and open map so we don't see errors
window.alert = jest.fn();
window.open = jest.fn();

const labelValueToolbarItem = (
  <LabelValueToolbarItem
    key="labelvalue"
    labelValue=""
    tooltip="Hello label"
    label="label"
    menuLabel="menu label"
    widthPx={50}
  />
);

const loadingSpinnerToolbarItem = (
  <LoadingSpinnerToolbarItem
    key="loadingspinner"
    itemsToLoad={1}
    tooltip="Hello LoadingSpinner"
    hasIssue={false}
    widthPx={50}
  />
);

describe('LabelValueOverflowMenuToolbarItem', () => {
  test('LabelValueOverflowMenuToolbarItem renders', () => {
    const { container } = render(
      <LabelValueOverflowMenuToolbarItem
        item={labelValueToolbarItem.props}
        menuKey={labelValueToolbarItem.key}
      />
    );

    expect(container).toMatchSnapshot();
  });
  test('LabelValueOverflowMenuToolbarItem renders empty due to incorrect type', () => {
    const { container } = render(
      <LabelValueOverflowMenuToolbarItem
        item={loadingSpinnerToolbarItem.props}
        menuKey={loadingSpinnerToolbarItem.key}
      />
    );

    expect(container).toMatchSnapshot();
  });
});
