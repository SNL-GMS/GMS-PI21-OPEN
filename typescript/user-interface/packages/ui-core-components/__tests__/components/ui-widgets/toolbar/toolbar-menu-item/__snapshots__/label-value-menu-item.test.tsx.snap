// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LabelValueOverflowMenuToolbarItem LabelValueOverflowMenuToolbarItem renders 1`] = `
<div>
  <li
    class=""
  >
    <a
      class="bp4-menu-item bp4-popover-dismiss"
      tabindex="0"
      title="Hello label"
    >
      <div
        class="bp4-fill bp4-text-overflow-ellipsis"
      >
        label:  
      </div>
    </a>
  </li>
</div>
`;

exports[`LabelValueOverflowMenuToolbarItem LabelValueOverflowMenuToolbarItem renders empty due to incorrect type 1`] = `<div />`;
